### ISSUE TITLE
collecting artifacts from different builds

### ISSUE THREAD

**Issue by stefanseefeld** [travis-ci/travis-ci#6224]
Question:

I already know how to collect build artifacts and how to deploy them.
What is the support for combining artifacts from multiple builds ?
I'm running tests for each build variant in my matrix, and want to compile a test report once all the builds have finished, then push that test report to my gh-pages branch.
Is there any support for that use-case ?

Thanks !

**Comment by BanzaiMan** [comment]
The artifacts addon does not have this functionality.

You can upload your build byproducts with the S3 deployment provider to a common location.

**Comment by stefanseefeld** [comment]
OK, but even with the "deploy" tool this isn't obvious. https://docs.travis-ci.com/user/deployment explains how to use a "file:" attribute, but it isn't clear what that is referring to when multiple builds are involved. Presumably the "deploy" is global (as opposed to local to individual builds in a build matrix), so how can I even refer to files produced by individual builds ?

**Comment by BanzaiMan** [comment]
The `file:` value in https://docs.travis-ci.com/user/deployment appears in an example for `releases`. It doesn't apply to the S3 provider I'm talking about. The S3 provider's documentation is found in https://docs.travis-ci.com/user/deployment/s3.

You can, for example, use the `upload-dir` value (https://docs.travis-ci.com/user/deployment/s3#Deploy-to-a-Specific-S3-Folder):

```yaml
provider: s3
⋮
  upload-dir: build/$TRAVIS_BUILD_NUMBER
```

**Comment by stefanseefeld** [comment]
But I do want to upload my test report to my github repo. How would an S3 provider help with that ?

**Comment by BanzaiMan** [comment]
My example is for the S3 provider. I'm not sure how you are getting hung up on unrelated GitHub Releases example.

**Comment by stefanseefeld** [comment]
We are talking past each other, it seems.
In my original message I was specifically asking for help in combining test results from multiple builds into a single test report that I could push to a gh-pages branch of my github repo.
If an S3 provider doesn't help me with that, I'm not sure why it even came up in the conversation.

**Comment by BanzaiMan** [comment]
Indeed, it seems that way.

Do you mean multiple "builds", or multiple "jobs"? In other words, are you talking about multiple builds https://travis-ci.org/OWNER/REPO/builds/X and https://travis-ci.org/OWNER/REPO/builds/Y? Or jobs within a build matrix?

My impression was that you meant the latter, and mentioned S3 as an alternative, since it affords you flexibility to define where files are uploaded.

And when you say: "compile a test report once all the builds have finished, then push that test report to my gh-pages branch", what does that really mean? What report? What writes this report? Are you trying to commit this report to the `gh-pages` branch (is it in the same repo, or a different one) and push it back to GitHub?

**Comment by stefanseefeld** [comment]
Yes, I think I'm talking about "jobs". (Though I was also considering collecting reports from other CI services, such as appveyor, to have better coverage.)
The idea with a "compiled" report is perhaps a little too ambitious, at least for now, since it involves more than just uploading build artifacts. (It could be done if I push the artifacts into a different branch or repo, and use the "push" trigger there to do a separate CI job just to compile the report.

Committing data to either the gh-pages branch or an entirely separate repo is the easy part. The open question (for now) is how to get hold of files produced in different jobs (associated with the same build number) and move them to a common place.

**Comment by BanzaiMan** [comment]
> The open question (for now) is how to get hold of files produced in different jobs (associated with the same build number) and move them to a common place.

You can deploy your build artifacts to a common location using S3.

**Comment by stefanseefeld** [comment]
So what would I need to put into my .travis.yml file to collect multiple files (one per build job, for the sake of simplicity) and collectively push them into a github repo  ?

**Comment by BanzaiMan** [comment]
I'm a bit confused.

You said you want to "move them to a common place". For that, you can use the S3 deployment to upload files to S3. See https://docs.travis-ci.com/user/deployment/s3#Deploy-to-a-Specific-S3-Folder.

If this is the case, ensure that the upload happens from each of the jobs within your build matrix. Then, take care to not have multiple files from different jobs going to the same path (e.g., `path/to/file1` from jobs 1 and 2 in the build matrix), since if deployment happens simultaneously to the identical path, the file might become corrupt.

If you want to add them to your git repository, you can use the `script` provider https://docs.travis-ci.com/user/deployment/script to set up git config (which can be skipped if you are dealing with the same repository that you are building from), then push to GitHub after setting up credentials (or create a commit via API https://developer.github.com/v3/git/commits/#create-a-commit).

I don't have good examples for these, as I don't think these are common.

If you have a repository to work with, I can make further suggestions. Thus far, the discussions have been quite abstract, and this is the best I can tell you with the information I have.

**Comment by stefanseefeld** [comment]
The repo for which I'm considering this is https://github.com/boostorg/python, and the relevant .travis.yml file is here: https://github.com/boostorg/python/blob/develop/.travis.yml

As you can see, there currently are 4 configs, each generating a test report. If I understand you correctly, such a `scripts/deploy.sh` script would be executed once for each job, so all I'd need to do is do a `git commit` each time. Correct ?
Is there a way to add an action that's only executed once per build, after all jobs have completed ? (That's where I could add the test report compilation that would combine the test reports from all jobs.)

Thanks,

**Comment by BanzaiMan** [comment]
Maybe.

It is probably OK, but it depends on what the commits would look like.

If each of these jobs are touching common files, you may have a merge conflict and may not be able to run `git push` successfully. In addition, if you run more than one builds and corresponding jobs from different builds end up running simultaneously, you might also have a conflict. You can limit the number of jobs this repository runs to 1 (https://docs.travis-ci.com/user/customizing-the-build/#Limiting-Concurrent-Builds), but you may not want to do that.

One task within a build is not currently supported. https://github.com/travis-ci/travis-ci/issues/929#issuecomment-35027109 shows a potential workaround, but, for now, caveat emptor applies.

**Comment by VoR0220** [comment]
I too am looking for something similar. I'm helping design a programming language which uses Travis and we're trying to configure the CI to take a handful of artifacts, push them up to a repo somewhere (ideally these artifacts could at some point be `shared` amongst the jobs and be collected via a mutex of some kind), where then we compare the artifacts to see if they are equal amongst multiple platforms (Cpp project running emscripten, os x, linux, windows, and docker...yea it's a lot...I have a feeling I'm going to need a hack of some kind)...my question is do I need to worry about builds timing out wrt to jobs since this would be something of an "after_script" job?

**Comment by hbons** [comment]
I've been trying to find how to do this as well. I want to create nightly builds for different platforms and upload them all to the same place (in my case GitHub Releases), but this is impossible without tricky hacks. I'm not sure what the proper solution would be, but being able to mark certain files in different jobs on which a common deploy script could operate would be really useful.

**Comment by ELLIOTTCABLE** [comment]
Not stale. Still a valid question / feature-request.

In my case, each of my tests in the matrix, in the process of testing, builds a binary artifact; and I wish to collect these per-matrix-job artifacts in a way that they're accessible to an eventual deploy-step.

As it stands, only the artifacts from the *specific* build-job preceding the deploy-step are still available, it seems?

**Comment by Gjacquenot** [comment]
Is the current solution is not to use parallel jobs and do everything sequentially?

**Comment by ELLIOTTCABLE** [comment]
@Gjacquenot I couldn't quite parse your response, but — forcing sequentiality doesn't solve this Issue; build-artifacts are still not stored centrally during Jobs in a useful way. @hbons basically nailed what we *need* here — something like the inverse of the build-cache; the ability to have each worker download-and-then-upload a “completed-artifact cache”, that the deploy worker can then refer to if all of said build-jobs succeed.

Closest solution right now is a manual implementation of exactly that: have each job's worker upload its artifacts to some central location (a dedicated “temporary artifact cache” S3 bucket, for instance); then have the final deploy worker or whatever you're using re-download the artifacts from there when/if it needs them.

**Comment by Gjacquenot** [comment]
I have found a solution : since each job calls the deploy part if the .travis.yml , one has to write a generic deploy section that can handle the names of the artifacts generated in parallel jobs. I use regex to match the names of thé generated files. The constraint is to give some kind of prefix to the artifacts.

Here is an example where I generate two artifacts in two parallel jobs that I send to github releases. 


https://github.com/Gjacquenot/ssc/blob/master/.travis.yml

**Comment by kousu** [comment]
I have this use case too: I am working on a [Stata](https://www.stata.com/) project: https://github.com/kousu/statasvm. Stata's packages combine all platforms into one folder, and then at install time it decides which files to actually download depending on your platform.

I was doing my builds using virtual machines with a _shared samba folder_:

1. check my repo out
2. mount the shared folder in all my VMs
3. manually run `make` in each VM in turn
  -💡 my Makefile arranges to put the platform-specific stuff into platform-specific subfolders, just like @Gjacquenot's example.
4. manually zip the final folder

I am keen to automate this because we've had several glitches between source and published package. I was considering building my own set of VMs and spawning them my own way (with qemu? qemu would work) but Travis already offers [Windows](https://docs.travis-ci.com/user/reference/windows/) *and* [macOS](https://docs.travis-ci.com/user/reference/osx/) in one place so I figure I will give it a shot. Because Stata is unusual (i.e. pre-web, pre-browser wars, pre-cross-platform) it doesn't follow the modern convention, where a single source repo generates multiple variant packages; instead I need a single source repo to generate a single package containing all the variants, so I really need a way to recombine artifacts from multiple parallel jobs at the end, and there is *no way* to just run the builds in parallel because I genuinely need entirely different OSes to do my builds on.

I imagine the IoT people are in a similar situation: they need to generate a hundred different artifacts in one location.
